### PR TITLE
Feature recrypt, refs: #593

### PR DESCRIPTION
**Pull request by luca-bruegger**
So far done. Needs merge (rebase) with master and another review from @mtnstar regarding the frontend changes for the recrypt.

### CODE REVIEW

File: app/utils/crypto/symmetric/recrypt.rb [hunk 1]
```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+class Crypto::Symmetric::Recrypt
+
+  def initialize(current_user, team, private_key)
+    @current_user = current_user
+    @team = team
+    @private_key = private_key
+
+    perform
```
Comment (mtnstar) [inline app/utils/crypto/symmetric/recrypt.rb:10]: call perform from the outside, not inside constructor ...
BlaClass.new(args).perform

File: app/utils/crypto/symmetric/recrypt.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  end
+
+  private
+
+  def perform
+    return if Crypto::EncryptionAlgorithm.latest_in_use?(@team)
```
Comment (mtnstar) [inline app/utils/crypto/symmetric/recrypt.rb:16]: don't we also have to check if recrypt is in progress or failed?

File: app/utils/crypto/symmetric/recrypt.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    member.password = encrypted_team_password
+    member.save!
+  end
+
+  def update_team_encryption_algorithm
+    @team.update_encryption_algorithm
+    @team.save!
+  end
+
+  def entailed_encryptables
```
Comment (mtnstar) [inline app/utils/crypto/symmetric/recrypt.rb:65]:
```suggestion
  def team_encryptables
```
?

File: db/migrate/20220517131300_change_tables_to_support_recrypt.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+class ChangeTablesToSupportRecrypt < ActiveRecord::Migration[7.0]
+  def change
+    add_column :teams, :recrypt_state, :integer, default: 1, null: false
+    add_column :teams, :encryption_algorithm, :string, default: 'AES256', null: false
```
Comment (mtnstar) [inline db/migrate/20220517131300_change_tables_to_support_recrypt.rb:4]: is it wise to use this deprecated value as default?
maybe first add new column ... update all existing entries .. then make column not null

File: app/utils/crypto/symmetric/recrypt.rb [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    @team.folders.map(&:encryptables).flatten
```
Comment (mtnstar) [inline app/utils/crypto/symmetric/recrypt.rb:66]: do not use map here.
use arel/sql to get all encryptables belonging to a team.
consider moving this getter to team model class
don't forget to include encryptable files which do not have a folder id assigned.

File: spec/unit/utils/crypto/symmetric/recrypt_spec.rb [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+    team = Team::Shared.create(alice, name: 'Team 1')
+
+    expect(team.encryption_algorithm).to eq 'AES256IV'
+
+    private_key = alice.decrypt_private_key('password')
+    described_class.new(alice, team, private_key)
+
+    expect(team.encryption_algorithm).to eq 'AES256IV'
+    expect(team.recrypt_state).to eq 'done'
```
Comment (mtnstar) [inline spec/unit/utils/crypto/symmetric/recrypt_spec.rb:21]: also test resetting of team password?

File: spec/unit/utils/crypto/symmetric/recrypt_spec.rb [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+  it 'recrypts given team' do
+    expect(Crypto::EncryptionAlgorithm).to receive(:latest).and_return(:AES256)
+
+    team = Team::Shared.create(alice, name: 'Team 1')
+
+    expect(team.encryption_algorithm).to eq 'AES256IV'
+
+    private_key = alice.decrypt_private_key('password')
+    described_class.new(alice, team, private_key)
```
Comment (mtnstar) [inline spec/unit/utils/crypto/symmetric/recrypt_spec.rb:18]: it's not obvious that new also performs the recrypt.
so as explained further up ... make sure you call perform explicit

File: spec/unit/utils/crypto/symmetric/recrypt_spec.rb [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+  it 'recrypts team encryptables' do
+    expect(Crypto::EncryptionAlgorithm).to receive(:latest).and_return(:AES256)
+    team = teams(:team1)
+
+    private_key = alice.decrypt_private_key('password')
+    described_class.new(alice, team, private_key)
+
+    encryptable_types =  team.folders.pluck(:encryptables)
+    require 'pry'; binding.pry
```
Comment (mtnstar) [inline spec/unit/utils/crypto/symmetric/recrypt_spec.rb:54]: pry ...

File: spec/unit/utils/crypto/symmetric/recrypt_spec.rb [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    expect(team.encryption_algorithm).to eq 'AES256IV'
+    expect(team.recrypt_state).to eq 'done'
+  end
+
+  private
+
+  def mocked_team
+    defect_encryptable = teams(:team1).folders.first.encryptables.first
+    defect_encryptable.encrypted_data.[]=(:password, **{iv: nil, data: 'encrypted'})
+    defect_encryptable.save
```
Comment (mtnstar) [inline spec/unit/utils/crypto/symmetric/recrypt_spec.rb:65]: invalid or broken instead of defect? defective ?

**Review by mtnstar — CHANGES_REQUESTED** [review]
so far so good :rocket: 

still some things to check:

- make sure you handle cases where previous recrypt has failed or recrypt is in progress
- extend specs to make sure every thing is tested .. including team password reset during recrypt, ...
- just keep it simple and use hash map for a list of available encryption algorithms

File: app/utils/crypto/encryption_algorithm.rb [hunk 10]
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+class Crypto::EncryptionAlgorithm
```
Comment (mtnstar) [inline app/utils/crypto/encryption_algorithm.rb:3]: can't we just add this to Crypto::Symmetric ?
does it make sense to have it's own class here?

File: app/models/team.rb [hunk 11]
```diff
@@ -60,6 +70,29 @@ def self.policy_class
     TeamPolicy
   end
 
+  def password_bytesize
+    Crypto::EncryptionAlgorithm.get_class(self.encryption_algorithm).password_bytesize
```
Comment (mtnstar) [inline app/models/team.rb:74]:
```suggestion
    Crypto::EncryptionAlgorithm::ALGORITHMS[self.encryption_algorithm].password_bytesize ?
```

File: app/models/team.rb [hunk 12]
```diff
@@ -60,6 +70,29 @@ def self.policy_class
...
+    return if Crypto::EncryptionAlgorithm.all.exclude?(algorithm)
+
+    write_attribute(:encryption_algorithm, algorithm)
+  end
+
+  def update_encryption_algorithm
+    self.encryption_algorithm = Crypto::EncryptionAlgorithm.latest
+  end
+
+  def encryptables
```
Comment (mtnstar) [inline app/models/team.rb:92]: for getting a list of all encryptables ?

File: app/utils/crypto/symmetric/aes256.rb [hunk 13]
```diff
@@ -12,25 +12,29 @@ class Crypto::Symmetric::Aes256 < Crypto::Symmetric
   ITERATION_COUNT ||= 1000
 
   class << self
+    password_bytesize = 32
```
Comment (mtnstar) [inline app/utils/crypto/symmetric/aes256.rb:15]: maybe use class_attribute in base class ?

File: app/utils/crypto/symmetric/recrypt.rb [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+class Crypto::Symmetric::Recrypt
+
+  def initialize(current_user, team, private_key)
+    @current_user = current_user
+    @team = team
+    @private_key = private_key
+  end
+
+  def perform
+    return if latest_in_use? || recrypt_not_ready?
```
Comment (mtnstar) [inline app/utils/crypto/symmetric/recrypt.rb:12]: maybe find a better name?
latest? what does latest mean here?
**already_recrypted?**

or .... requires_recrypt? and invert with unless

**Review by mtnstar — CHANGES_REQUESTED** [review]
some more feedback ... keep going :wink:

File: spec/unit/utils/crypto/symmetric/recrypt_spec.rb [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+    expect(team.teammembers).should_not match(old_team_passwords)
+
+    expect(team.encryption_algorithm).to eq 'AES256IV'
+    expect(team.recrypt_state).to eq 'done'
+  end
+
+  private
+
+  def mocked_team
```
Comment (mtnstar) [inline spec/unit/utils/crypto/symmetric/recrypt_spec.rb:100]: create_broken_encryptable ?
why returning team fixture here?
maybe just create borken encryptable ... it's included in team1 encryptables then

File: app/models/team.rb [hunk 16]
```diff
@@ -60,6 +70,32 @@ def self.policy_class
     TeamPolicy
   end
 
+  def password_bytesize
+    Crypto::Symmetric::EncryptionAlgorithm::ALGORITHMS[encryption_algorithm].password_bitsize
```
[the remark this answers is not in the document]
Comment (luca-bruegger) [reply app/models/team.rb:74]: I assume that bitsize is better than bytesize. As 256 stands for bits and not bytes.

File: app/models/team.rb [hunk 17]
```diff
@@ -24,6 +24,12 @@ class Team < ApplicationRecord
   validates :name, length: { maximum: 40 }
   validates :description, length: { maximum: 300 }
```
Comment (mtnstar) [inline app/models/team.rb:25]: add something like:

`validates :encryption_algorithm, inclusion: Crypto::Symmetric::ALGORITHMS.keys`

File: app/models/team.rb [hunk 18]
```diff
@@ -60,10 +70,39 @@ def self.policy_class
...
+  def password_bytesize
+    Crypto::Symmetric::ALGORITHMS[encryption_algorithm].password_bitsize
+  end
+
+  def encryption_algorithm
+    # using read_attribute to avoid recursion
+    read_attribute(:encryption_algorithm) || Crypto::Symmetric.latest_algorithm
+  end
+
+  def encryption_algorithm=(algorithm)
```
Comment (mtnstar) [inline app/models/team.rb:82]: use validation inclusion instead of overriding setter

File: app/models/team.rb [hunk 19]
```diff
@@ -60,10 +70,39 @@ def self.policy_class
     TeamPolicy
   end
 
+  def password_bytesize
+    Crypto::Symmetric::ALGORITHMS[encryption_algorithm].password_bitsize
```
Comment (mtnstar) [inline app/models/team.rb:74]: create encryption_class getter method and use it also here

File: app/models/team.rb [hunk 20]
```diff
@@ -60,10 +70,39 @@ def self.policy_class
...
+    return if Crypto::Symmetric.all.exclude?(algorithm)
+
+    write_attribute(:encryption_algorithm, algorithm)
+  end
+
+  def new_team_password
+    Crypto::Symmetric::ALGORITHMS[encryption_algorithm].random_key
+  end
+
+  def update_encryption_algorithm
```
Comment (mtnstar) [inline app/models/team.rb:92]: set this from outside in recrypt domain class
team model should not care about this

File: app/utils/crypto/symmetric.rb [hunk 21]
```diff
@@ -18,5 +26,17 @@ def decrypt
     def random_key
       raise 'Implement in subclass'
     end
+
+    def latest_algorithm
```
Comment (mtnstar) [inline app/utils/crypto/symmetric.rb:30]: maybe add as class constant at top?

`LATEST_ALGORITHM = 'AES256'`

File: app/utils/crypto/symmetric.rb [hunk 22]
```diff
@@ -18,5 +26,17 @@ def decrypt
...
+      'AES256'
+    end
+
+    def all
```
Comment (mtnstar) [inline app/utils/crypto/symmetric.rb:34]: naming? symmetric.all?

File: app/utils/crypto/symmetric/recrypt.rb [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+    @private_key = private_key
+  end
+
+  def perform
+    return if already_recrypted? || recrypt_locked?
+
+    prepare_recrypt
+    begin
+      ActiveRecord::Base.transaction do
+
```
Comment (mtnstar) [inline app/utils/crypto/symmetric/recrypt.rb:17]: first set new algorithm on team

File: db/migrate/20220517131300_change_tables_to_support_recrypt.rb [hunk 24]
```diff
@@ -0,0 +1,17 @@
...
+
+    add_column :encryptables, :encryption_algorithm, :string
+
+    set_default_encryption_algorithm
+  end
+
+  private
+
+  def set_default_encryption_algorithm
+    Team.update_all(encryption_algorithm: "AES256")
```
Comment (mtnstar) [inline db/migrate/20220517131300_change_tables_to_support_recrypt.rb:14]:
```suggestion
    Team.update_all(encryption_algorithm: 'AES256')
```

File: db/migrate/20220517131300_change_tables_to_support_recrypt.rb [hunk 25]
```diff
@@ -0,0 +1,17 @@
+class ChangeTablesToSupportRecrypt < ActiveRecord::Migration[7.0]
+  def change
+    add_column :teams, :recrypt_state, :integer, default: 1, null: false
```
Comment (mtnstar) [inline db/migrate/20220517131300_change_tables_to_support_recrypt.rb:3]: do not use magic number 1 here ...
read it from team recrypt state

File: db/migrate/20220609063817_rename_password_to_encrypted_team_password.rb [hunk 26]
```diff
@@ -0,0 +1,5 @@
+class RenamePasswordToEncryptedTeamPassword < ActiveRecord::Migration[7.0]
+  def change
+    rename_column :teammembers, :password, :encrypted_team_password
```
Comment (mtnstar) [inline db/migrate/20220609063817_rename_password_to_encrypted_team_password.rb:3]: :heart_eyes:

File: spec/unit/utils/crypto/symmetric/recrypt_spec.rb [hunk 27]
```diff
@@ -0,0 +1,101 @@
...
+    expect(team.teammembers).not_to eq(old_team_passwords)
+
+    expect(team.encryption_algorithm).to eq 'AES256IV'
+    expect(team.recrypt_state).to eq 'done'
+  end
+
+  private
+
+  def create_broken_encryptable(team)
+    broken_encryptable = team.folders.first.encryptables.first
```
Comment (mtnstar) [inline spec/unit/utils/crypto/symmetric/recrypt_spec.rb:96]:
```suggestion
    broken_encryptable = team.encryptables.first
```

File: spec/unit/utils/crypto/symmetric/recrypt_spec.rb [hunk 28]
```diff
@@ -0,0 +1,101 @@
...
+
+    expect(team.encryption_algorithm).to eq 'AES256'
+
+    private_key = alice.decrypt_private_key('password')
+    described_class.new(alice, team, private_key).perform
+
+    expect(team.encryption_algorithm).to eq 'AES256IV'
+    expect(team.recrypt_state).to eq 'done'
+  end
+
```
Comment (mtnstar) [inline spec/unit/utils/crypto/symmetric/recrypt_spec.rb:22]: make also sure Encryptable::Files are also recrypted !

File: app/models/team.rb [hunk 29]
```diff
@@ -60,10 +70,39 @@ def self.policy_class
...
+    self.encryption_algorithm = Crypto::Symmetric.latest_algorithm
+  end
+
+  def encryptables
+    Encryptable.joins("LEFT JOIN folders ON encryptables.folder_id = folders.id
```
[the remark this answers is not in the document]
Comment (mtnstar) [reply app/models/team.rb:97]: do not forget to include encryptable files !

**Review by mtnstar — CHANGES_REQUESTED** [review]
thanks again for the great effort getting this done ! good work.

check the comments ... some important inputs and maybe things we should discuss ... anyway ... looks like reaching the :checkered_flag: soon :wink:

File: app/models/encryptable.rb [hunk 30]
```diff
@@ -27,14 +27,24 @@ class Encryptable < ApplicationRecord
...
-  def encrypt(_team_password)
+  def encrypt(_team_password, _algorithm = team.encryption_algorithm)
     raise 'implement in subclass'
   end
 
   def decrypt(_team_password)
     raise 'implement in subclass'
   end
 
+  def recrypt(team_password, new_team_password)
```
Comment (mtnstar) [inline app/models/encryptable.rb:38]: maybe add param current_encryption_algorithm ?

**Review by mtnstar — CHANGES_REQUESTED** [review]

File: app/models/team/shared.rb [hunk 31]
```diff
@@ -29,6 +29,10 @@ def create(creator, params)
       end
       team
     end
+
+    def policy_class
+      TeamPolicy
+    end
```
Comment (luca-bruegger) [inline app/models/team/shared.rb:35]: Move into self needed?

File: db/migrate/20220517131300_change_tables_to_support_recrypt.rb [hunk 32]
```diff
@@ -0,0 +1,14 @@
...
+    add_column :teams, :encryption_algorithm, :string
+
+    set_default_encryption_algorithm
+  end
+
+  private
+
+  def set_default_encryption_algorithm
+    Team.update_all(encryption_algorithm: 'AES256')
+  end
```
Comment (luca-bruegger) [inline db/migrate/20220517131300_change_tables_to_support_recrypt.rb:13]: Maybe remove method and directly call `update_all`?

File: spec/migrations/use_encrypted_data_for_account_credentials_spec.rb.disabled [hunk 33]
```diff
@@ -193,10 +194,12 @@ def decrypt(team_password)
   private
 
   def decrypt_attr(attr, team_password)
-    encrypted_value = encrypted_data[attr].try(:[], :data)
+    data = encrypted_data[attr].try(:[], :data)
+    iv = encrypted_data[attr].try(:[], :iv)
 
-    cleartext_value = if encrypted_value
-                        Crypto::Symmetric::Aes256.decrypt(encrypted_value, team_password)
+    encrypted_values = { data: data, iv: iv }
```
Comment (luca-bruegger) [inline spec/migrations/use_encrypted_data_for_account_credentials_spec.rb.disabled:200]: Create additional method to retrieve `encrypted_values` to increase readability and to keep it DRY?

**Review by luca-bruegger — COMMENTED** [review]
Some little inputs

File: spec/migrations/use_encrypted_data_for_account_credentials_spec.rb.disabled (hunk 33 above)
Comment (luca-bruegger) [reply spec/migrations/use_encrypted_data_for_account_credentials_spec.rb.disabled:200]: Ignored as migration is disabled

File: app/models/team.rb [hunk 34]
```diff
@@ -60,10 +70,31 @@ def self.policy_class
     TeamPolicy
   end
 
+  def encryption_class
+    Crypto::Symmetric::ALGORITHMS[encryption_algorithm]
+  end
+
+  def password_bytesize
```
Comment (mtnstar) [inline app/models/team.rb:77]: byte vs. bit ?

File: spec/unit/utils/crypto/symmetric/recrypt_spec.rb [hunk 35]
```diff
@@ -0,0 +1,122 @@
...
+    stub_const('::Crypto::Symmetric::LATEST_ALGORITHM', 'AES256')
+
+    create_broken_encryptable(team)
+
+    stub_const('::Crypto::Symmetric::LATEST_ALGORITHM', 'AES256IV')
+
+    expect do
+      described_class.new(admin, team, admin_pk).perform
+    end.to raise_error(RuntimeError, 'Recrypt failed: wrong final block length')
+
```
Comment (mtnstar) [inline spec/unit/utils/crypto/symmetric/recrypt_spec.rb:72]: also here .. add check that encryptable can still be decrypted as before

**Review by mtnstar — APPROVED** [review]
looks almost perfect now !!! :rocket: :partying_face: 

just check the comments ... small improvements ...
we wanna make sure nothing goes wrong with this important feature

**Comment by mtnstar** [comment]
going to be merged with #728

**Pull request closed without merging**